pubsub/gossipsub/gossipsub-v1.4.md

```diff
@@ -0,0 +1,164 @@
+| `preamble_threshold` | The smallest message size to use message preamble | 200KB???  |
+
+
+
+## Message Preamble
+
+### Basic scenario
+
+When a peer starts relaying a message that exceeds the preamble_threshold size, 
+it may include a control message (called message preamble) at the start of the message.
+
+The purpose of the preamble is to allow receivers to instantly learn about the incoming message. 
+The preamble must include the message ID and length, 
+providing receivers with immediate access to critical information about the incoming message. 
+The receiver must immediately process a message preamble without waiting to download the entire message.
+
+On receiving a preamble that advertises a message ID not present in the seen cache, 
+a receiver should add that message ID to the ongoing_receives list.
+The ongoing_receives list is crucial in limiting IWANT requests and simultaneous reception of the same message from mesh peers.
+ 
+The receiver may use message length to leniently estimate message download time.
+If the message is successfully downloaded before the estimated download time has elapsed, 
+the message ID is removed from the ongoing_receives list and added to the seen cache. 
+
+If the download takes longer than the estimated download time, 
+the sender may be penalized through P4 to discourage peers from intentionally delaying message transmission.
+
+The message preamble is considered _optional_ for both the sender and the receiver. 
+This means that the sender may choose not to prepend the preamble, 
+and the receiver may also opt to ignore it.
+Adding a message preamble may increase control overhead for small messages. 
+Therefore, it is preferable to use it only for messages that exceed the preamble_threshold.
+
+
+### Limiting IWANT Requests
+
+When a peer receives an IHAVE announcement for a message ID not present in the seen cache, 
+the peer must also check the  ongoing_receives list before making an IWANT request. 
+
+If the message ID is found in the ongoing_receives list, 
+the peer should postpone sending the IWANT request for a defer_interval. 
+The defer_interval may be based on the message download time.
+
+If the message download completes before the defer_interval expires, 
+the IWANT request for that message will not be generated. 
+However, if the defer_interval elapses and the download has not completed, 
+the peer can proceed to make the IWANT request for the missing message. 
+
+The total number of outstanding IWANT requests for a single message should not exceed max_iwant_requests.
+Every peer must respond to the received IWANT requests. 
+Failing to do so may result in behavioural penalty through P7. 
+This will discourage peers from intentionally not replying to IWANT requests.
+
+### IMReceiving Message
+
+The IMReceiving message serves a distinct purpose compared to the IDONTWANT message. 
+An IDONTWANT can only be transmitted after receiving the entire message. 

```

Comment:
While I understand the distinction here of `IMReceiving` compared with `IDONTWANT` and having this broadcasted earlier, how effective would this be in practice ? One issue we have seen is that an actual control message takes a while to be processed by the gossip router even after it has been received due to HOL blocking. So by the time you process the control message, the actual message might already be sent by your mesh peers.

Reply:
> For the scenarios we tested, sending IMReceiving significantly increases the probability of mesh peers being able to stop unnecessary message sends since enough IMReceiving go through in time.

Is there more information on the scenarios tested ? Ex: How many different topics nodes were subscribed to along with how many messages were being published per second on these topics.

Reply:
> Is there more information on the scenarios tested ? Ex: How many different topics nodes were subscribed to along with how many messages were being published per second on these topics.

All (1500) peers were subscribed to a single topic. Twelve messages were introduced, each by a different publisher, with each publisher waiting 3 seconds before sending the next message. Messages larger than 600 KB take more time to reach all peers, building outgoing message queues at many peers.

Reply:
In the results below, we consider 1,500 peers (single topic) with an inter-message spacing of 50 ms, which is roughly 20 messages per second. The message size is 50 KB.
![S3 lat BW graph](https://github.com/user-attachments/assets/1c788b45-497c-44fb-a593-f55ab984f4fb)